File: packages/ingest/src/lib/usc/cross-references.ts
type QualifierType =
	| "subsection"
	| "subdivision"
	| "paragraph"
	| "subparagraph"
	| "clause";

interface Qualifier {
	type: QualifierType;
	designators: string[];
}

type SectionMention = {
	section: string;
	titleNum: string | null;
	offset: number;
	length: number;
};

type SectionTarget =
	| {
			type: "section";
			mention: SectionMention;
	  }
	| {
			type: "range";
			start: SectionMention;
			end: SectionMention;
			inclusive: boolean;
	  };

export interface SectionCrossReference {
	section: string;
	titleNum: string | null;
	offset: number;
	length: number;
	link: string | null;
}

type Token =
	| {
			type: "sectionNumber";
			value: string;
			titleNum: string | null;
			start: number;
			end: number;
	  }
	| { type: "titleNumber"; value: string; start: number; end: number }
	| { type: "designator"; value: string }
	| { type: "word"; value: string }
	| { type: "punct"; value: "," | ";" | "." | ":" };

// USC section numbers are simpler: digits with optional letter suffix (e.g., "1234", "1234a", "5")
const SECTION_NUMBER_RE = /^\d+[a-zA-Z]*(?:-\d+)?$/;
// Title number reference
const TITLE_NUMBER_RE = /^\d+$/;
const DESIGNATOR_RE = /^\(([A-Za-z0-9ivxIVX]+)\)$/;
// Token regex: matches section numbers, designators, words (including U.S.C.), and punctuation
const TOKEN_RE =
	/\d+[a-zA-Z]*(?:-\d+)?|\([A-Za-z0-9ivxIVX]+\)|U\.?S\.?C\.?|[A-Za-z]+(?:\/[A-Za-z]+)?|[,.;:§]/g;

const QUALIFIER_KEYWORDS = new Map<string, QualifierType>([
	["subsection", "subsection"],
	["subsections", "subsection"],
	["subdivision", "subdivision"],
	["subdivisions", "subdivision"],
	["paragraph", "paragraph"],
	["paragraphs", "paragraph"],
	["subparagraph", "subparagraph"],
	["subparagraphs", "subparagraph"],
	["clause", "clause"],
	["clauses", "clause"],
]);

const SECTION_KEYWORDS = new Set(["section", "sections", "sec", "secs"]);
const TITLE_KEYWORDS = new Set(["title"]);
const USC_KEYWORDS = new Set(["usc", "u.s.c.", "u.s.c"]);
const SEPARATOR_WORDS = new Set(["and", "or", "and/or"]);

/**
 * Extract cross-references from USC section text.
 * @param text The section body text
 * @param currentTitleNum The title number of the current section (for relative references)
 */
export function extractSectionCrossReferences(
	text: string,
	currentTitleNum: string,
): SectionCrossReference[] {
	const tokens = tokenize(text);
	const references: SectionCrossReference[] = [];

	let index = 0;
	while (index < tokens.length) {
		const token = tokens[index];
		if (!token) {
			index += 1;
			continue;
		}

		// Check for "42 U.S.C. 1234" style references
		if (token.type === "titleNumber") {
			const parsed = parseTitleUSCReference(tokens, index);
			if (parsed) {
				references.push(...parsed.references);
				index = parsed.nextIndex;
				continue;
			}
		}

		// Check for "section 1234" or "section 1234 of title 42" style references
		if (isQualifierKeyword(token) || isSectionKeyword(token)) {
			const parsed = parseReference(tokens, index, currentTitleNum);
			if (parsed) {
				references.push(...parsed.references);
				index = parsed.nextIndex;
				continue;
			}
		}

		index += 1;
	}

	return dedupeReferences(references);
}

function tokenize(text: string): Token[] {
	const tokens: Token[] = [];
	const matches = text.matchAll(TOKEN_RE);

	for (const match of matches) {
		const raw = match[0];
		if (!raw) continue;

		const start = match.index ?? 0;
		const end = start + raw.length;

		// Check for section symbol
		if (raw === "§") {
			tokens.push({ type: "word", value: "section" });
			continue;
		}

		// Check for U.S.C. pattern
		if (USC_KEYWORDS.has(raw.toLowerCase().replace(/\./g, ""))) {
			tokens.push({ type: "word", value: "usc" });
			continue;
		}

		// Check if it's a number (could be title or section number)
		if (TITLE_NUMBER_RE.test(raw)) {
			tokens.push({
				type: "titleNumber",
				value: raw,
				start,
				end,
			});
			continue;
		}

		// Check for section number with letter suffix
		if (SECTION_NUMBER_RE.test(raw)) {
			tokens.push({
				type: "sectionNumber",
				value: normalizeSectionNumber(raw),
				titleNum: null,
				start,
				end,
			});
			continue;
		}

		const designatorMatch = raw.match(DESIGNATOR_RE);
		if (designatorMatch) {
			tokens.push({ type: "designator", value: designatorMatch[1] });
			continue;
		}

		if (raw === "," || raw === ";" || raw === "." || raw === ":") {
			tokens.push({ type: "punct", value: raw });
			continue;
		}

		tokens.push({ type: "word", value: raw.toLowerCase() });
	}

	return tokens;
}

function normalizeSectionNumber(value: string): string {
	return value.toLowerCase();
}

function isQualifierKeyword(
	token: Token | undefined,
): token is { type: "word"; value: string } {
	return token?.type === "word" && QUALIFIER_KEYWORDS.has(token.value);
}

function isSectionKeyword(
	token: Token | undefined,
): token is { type: "word"; value: string } {
	return token?.type === "word" && SECTION_KEYWORDS.has(token.value);
}

function isTitleKeyword(
	token: Token | undefined,
): token is { type: "word"; value: string } {
	return token?.type === "word" && TITLE_KEYWORDS.has(token.value);
}

function isUSCKeyword(
	token: Token | undefined,
): token is { type: "word"; value: string } {
	return token?.type === "word" && token.value === "usc";
}

function isWord(token: Token | undefined, value: string): boolean {
	return token?.type === "word" && token.value === value;
}

function isDesignator(
	token: Token | undefined,
): token is { type: "designator"; value: string } {
	return token?.type === "designator";
}

function isSectionNumber(token: Token | undefined): token is {
	type: "sectionNumber";
	value: string;
	titleNum: string | null;
	start: number;
	end: number;
} {
	return token?.type === "sectionNumber";
}

function isTitleNumber(token: Token | undefined): token is {
	type: "titleNumber";
	value: string;
	start: number;
	end: number;
} {
	return token?.type === "titleNumber";
}

function isSeparator(token: Token | undefined): boolean {
	if (!token) return false;
	if (token.type === "punct") {
		return token.value === "," || token.value === ";";
	}
	return token.type === "word" && SEPARATOR_WORDS.has(token.value);
}

/**
 * Parse "42 U.S.C. 1234" style references
 */
function parseTitleUSCReference(
	tokens: Token[],
	startIndex: number,
): { references: SectionCrossReference[]; nextIndex: number } | null {
	const titleToken = tokens[startIndex];
	if (!isTitleNumber(titleToken)) return null;

	const uscToken = tokens[startIndex + 1];
	if (!uscToken || !isUSCKeyword(uscToken)) return null;

	// Now expect section number(s)
	const sectionList = parseSectionList(
		tokens,
		startIndex + 2,
		true,
		titleToken.value,
	);
	if (!sectionList) return null;

	const references = buildReferences(sectionList.items);
	return { references, nextIndex: sectionList.nextIndex };
}

function parseReference(
	tokens: Token[],
	startIndex: number,
	currentTitleNum: string,
): { references: SectionCrossReference[]; nextIndex: number } | null {
	const token = tokens[startIndex];
	if (!token) return null;

	if (isQualifierKeyword(token)) {
		const qualifierChains = parseQualifierChainList(tokens, startIndex);
		if (!qualifierChains) return null;

		let index = qualifierChains.nextIndex;
		if (!isWord(tokens[index], "of")) return null;
		index += 1;

		const sectionKeyword = tokens[index];
		if (!sectionKeyword || !isSectionKeyword(sectionKeyword)) return null;

		const allowMultiple =
			sectionKeyword.value === "sections" || sectionKeyword.value === "secs";
		const sectionList = parseSectionListWithTitle(
			tokens,
			index + 1,
			allowMultiple,
			currentTitleNum,
		);
		if (!sectionList) return null;

		const references = buildReferences(sectionList.items);
		return { references, nextIndex: sectionList.nextIndex };
	}

	if (isSectionKeyword(token)) {
		const sectionList = parseSectionListWithTitle(
			tokens,
			startIndex + 1,
			true,
			currentTitleNum,
		);
		if (!sectionList) return null;
		const references = buildReferences(sectionList.items);
		return { references, nextIndex: sectionList.nextIndex };
	}

	return null;
}

function parseQualifierChainList(
	tokens: Token[],
	startIndex: number,
): { chains: Qualifier[][]; nextIndex: number } | null {
	const firstChain = parseQualifierChain(tokens, startIndex);
	if (!firstChain) return null;

	const chains: Qualifier[][] = [firstChain.qualifiers];
	let index = firstChain.nextIndex;

	while (true) {
		const separatorIndex = consumeSeparators(tokens, index);
		if (separatorIndex === null) break;

		const nextToken = tokens[separatorIndex];
		if (!nextToken || !isQualifierKeyword(nextToken)) {
			break;
		}

		const nextChain = parseQualifierChain(tokens, separatorIndex);
		if (!nextChain) break;

		chains.push(nextChain.qualifiers);
		index = nextChain.nextIndex;
	}

	return { chains, nextIndex: index };
}

function parseQualifierChain(
	tokens: Token[],
	startIndex: number,
): { qualifiers: Qualifier[]; nextIndex: number } | null {
	const qualifier = parseQualifier(tokens, startIndex);
	if (!qualifier) return null;

	const qualifiers: Qualifier[] = [qualifier.qualifier];
	let index = qualifier.nextIndex;

	while (isWord(tokens[index], "of")) {
		const nextToken = tokens[index + 1];
		if (!nextToken || !isQualifierKeyword(nextToken)) break;

		const nextQualifier = parseQualifier(tokens, index + 1);
		if (!nextQualifier) break;

		qualifiers.push(nextQualifier.qualifier);
		index = nextQualifier.nextIndex;
	}

	return { qualifiers, nextIndex: index };
}

function parseQualifier(
	tokens: Token[],
	startIndex: number,
): { qualifier: Qualifier; nextIndex: number } | null {
	const token = tokens[startIndex];
	if (!token || !isQualifierKeyword(token)) return null;

	const type = QUALIFIER_KEYWORDS.get(token.value);
	if (!type) return null;

	const list = parseDesignatorList(tokens, startIndex + 1);
	if (!list) return null;

	return {
		qualifier: { type, designators: list.designators },
		nextIndex: list.nextIndex,
	};
}

function parseDesignatorList(
	tokens: Token[],
	startIndex: number,
): { designators: string[]; nextIndex: number } | null {
	const first = tokens[startIndex];
	if (!isDesignator(first)) return null;

	const designators: string[] = [first.value];
	let index = startIndex + 1;

	while (true) {
		const sepIndex = consumeSeparators(tokens, index);
		if (sepIndex === null) break;

		const nextToken = tokens[sepIndex];
		if (!isDesignator(nextToken)) break;

		designators.push(nextToken.value);
		index = sepIndex + 1;
	}

	return { designators, nextIndex: index };
}

/**
 * Parse section list, looking for optional "of title X" at the end
 */
function parseSectionListWithTitle(
	tokens: Token[],
	startIndex: number,
	allowMultiple: boolean,
	defaultTitleNum: string,
): { items: SectionTarget[]; nextIndex: number } | null {
	const sectionList = parseSectionList(
		tokens,
		startIndex,
		allowMultiple,
		defaultTitleNum,
	);
	if (!sectionList) return null;

	let index = sectionList.nextIndex;
	let titleNum = defaultTitleNum;

	// Check for "of title X" pattern
	if (isWord(tokens[index], "of") && isTitleKeyword(tokens[index + 1])) {
		const titleNumToken = tokens[index + 2];
		if (isTitleNumber(titleNumToken)) {
			titleNum = titleNumToken.value;
			index = index + 3;

			// Update all items with the explicit title number
			for (const item of sectionList.items) {
				if (item.type === "section") {
					item.mention.titleNum = titleNum;
				} else {
					item.start.titleNum = titleNum;
					item.end.titleNum = titleNum;
				}
			}
		}
	}

	return { items: sectionList.items, nextIndex: index };
}

function parseSectionList(
	tokens: Token[],
	startIndex: number,
	allowMultiple: boolean,
	defaultTitleNum: string | null,
): { items: SectionTarget[]; nextIndex: number } | null {
	const firstItem = parseSectionItem(tokens, startIndex, defaultTitleNum);
	if (!firstItem) return null;

	const items: SectionTarget[] = [firstItem.item];
	let index = firstItem.nextIndex;

	if (!allowMultiple) {
		return { items, nextIndex: index };
	}

	while (true) {
		const sepIndex = consumeSeparators(tokens, index);
		if (sepIndex === null) break;

		let nextIndex = sepIndex;
		if (isSectionKeyword(tokens[nextIndex])) {
			nextIndex += 1;
		}

		const nextItem = parseSectionItem(tokens, nextIndex, defaultTitleNum);
		if (!nextItem) break;

		items.push(nextItem.item);
		index = nextItem.nextIndex;
	}

	return { items, nextIndex: index };
}

function parseSectionItem(
	tokens: Token[],
	startIndex: number,
	defaultTitleNum: string | null,
): { item: SectionTarget; nextIndex: number } | null {
	let token = tokens[startIndex];

	// Handle case where titleNumber could actually be a section number
	if (isTitleNumber(token)) {
		token = {
			type: "sectionNumber",
			value: token.value,
			titleNum: defaultTitleNum,
			start: token.start,
			end: token.end,
		};
	}

	if (!isSectionNumber(token)) return null;

	let index = startIndex + 1;
	const startMention: SectionMention = {
		section: token.value,
		titleNum: token.titleNum ?? defaultTitleNum,
		offset: token.start,
		length: token.end - token.start,
	};

	if (isWord(tokens[index], "to") || isWord(tokens[index], "through")) {
		let endToken = tokens[index + 1];

		// Handle titleNumber as sectionNumber
		if (isTitleNumber(endToken)) {
			endToken = {
				type: "sectionNumber",
				value: endToken.value,
				titleNum: defaultTitleNum,
				start: endToken.start,
				end: endToken.end,
			};
		}

		if (!isSectionNumber(endToken)) return null;

		index += 2;
		let inclusive = false;

		if (tokens[index]?.type === "punct" && tokens[index]?.value === ",") {
			if (isWord(tokens[index + 1], "inclusive")) {
				inclusive = true;
				index += 2;
			}
		} else if (isWord(tokens[index], "inclusive")) {
			inclusive = true;
			index += 1;
		}

		const endMention: SectionMention = {
			section: endToken.value,
			titleNum: endToken.titleNum ?? defaultTitleNum,
			offset: endToken.start,
			length: endToken.end - endToken.start,
		};

		return {
			item: { type: "range", start: startMention, end: endMention, inclusive },
			nextIndex: index,
		};
	}

	return { item: { type: "section", mention: startMention }, nextIndex: index };
}

function consumeSeparators(tokens: Token[], startIndex: number): number | null {
	let index = startIndex;
	let consumed = false;

	while (isSeparator(tokens[index])) {
		consumed = true;
		index += 1;
	}

	return consumed ? index : null;
}

function buildReferences(items: SectionTarget[]): SectionCrossReference[] {
	const refs: SectionCrossReference[] = [];

	for (const item of items) {
		if (item.type === "section") {
			refs.push(buildReference(item.mention));
			continue;
		}

		refs.push(buildReference(item.start), buildReference(item.end));
	}

	return refs;
}

function dedupeReferences(
	references: SectionCrossReference[],
): SectionCrossReference[] {
	const seen = new Set<string>();
	const result: SectionCrossReference[] = [];

	for (const ref of references) {
		const key = `${ref.section}:${ref.titleNum}:${ref.offset}:${ref.length}:${ref.link}`;
		if (seen.has(key)) continue;
		seen.add(key);
		result.push(ref);
	}

	return result;
}

function buildReference(mention: SectionMention): SectionCrossReference {
	return {
		section: mention.section,
		titleNum: mention.titleNum,
		offset: mention.offset,
		length: mention.length,
		link: buildSectionLink(mention.section, mention.titleNum),
	};
}

function buildSectionLink(
	section: string,
	titleNum: string | null,
): string | null {
	if (!titleNum) return null;
	return `/statutes/usc/section/${titleNum}/${section}`;
}

File: packages/ingest/container-rust/src/sources/cgs/cross_references.rs
use regex::Regex;
use serde::{Deserialize, Serialize};
use std::collections::HashSet;
use std::sync::LazyLock;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct SectionCrossReference {
    pub section: String,
    pub offset: usize,
    pub length: usize,
    pub link: String,
}

#[derive(Debug, Clone)]
enum Token {
    SectionNumber {
        value: String,
        start: usize,
        end: usize,
    },
    Designator {
        #[allow(dead_code)]
        value: String,
    },
    Word {
        value: String,
    },
    Punct {
        value: char,
    },
}

#[derive(Debug, Clone)]
struct SectionMention {
    section: String,
    offset: usize,
    length: usize,
}

#[derive(Debug, Clone)]
enum SectionTarget {
    Section {
        mention: SectionMention,
    },
    Range {
        start: SectionMention,
        end: SectionMention,
        #[allow(dead_code)]
        inclusive: bool,
    },
}

static SECTION_NUMBER_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"^\d+[a-zA-Z]*-(?:\d+[a-zA-Z]*)(?:-\d+[a-zA-Z]*)*$").unwrap());
static DESIGNATOR_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"^\(([A-Za-z0-9ivxIVX]+)\)$").unwrap());
static TOKEN_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(
        r"\d+[a-zA-Z]*-(?:\d+[a-zA-Z]*)(?:-\d+[a-zA-Z]*)*|\([A-Za-z0-9ivxIVX]+\)|[A-Za-z]+(?:/[A-Za-z]+)?|[,.;:]",
    )
    .unwrap()
});

static QUALIFIER_KEYWORDS: LazyLock<HashSet<&str>> = LazyLock::new(|| {
    [
        "subsection",
        "subsections",
        "subdivision",
        "subdivisions",
        "paragraph",
        "paragraphs",
        "subparagraph",
        "subparagraphs",
        "clause",
        "clauses",
    ]
    .into_iter()
    .collect()
});

static SECTION_KEYWORDS: LazyLock<HashSet<&str>> =
    LazyLock::new(|| ["section", "sections", "sec", "secs"].into_iter().collect());
static SEPARATOR_WORDS: LazyLock<HashSet<&str>> =
    LazyLock::new(|| ["and", "or", "and/or"].into_iter().collect());

pub fn extract_section_cross_references(text: &str) -> Vec<SectionCrossReference> {
    let tokens = tokenize(text);
    let mut references = Vec::new();

    let mut index = 0;
    while index < tokens.len() {
        let token = &tokens[index];

        if is_qualifier_keyword(token) || is_section_keyword(token) {
            if let Some(parsed) = parse_reference(&tokens, index) {
                references.extend(parsed.references);
                index = parsed.next_index;
                continue;
            }
        }

        if is_section_number(token) {
            if let Some(section_list) = parse_section_list(&tokens, index, true) {
                references.extend(build_references(&section_list.items));
                index = section_list.next_index;
                continue;
            }
        }

        index += 1;
    }

    dedupe_references(references)
}

pub fn inline_section_cross_references(text: &str) -> String {
    let mut references = extract_section_cross_references(text);
    references.sort_by(|a, b| b.offset.cmp(&a.offset));

    let mut output = text.to_string();
    for reference in references {
        let start = reference.offset;
        let end = start.saturating_add(reference.length);
        if end > output.len()
            || start >= end
            || !output.is_char_boundary(start)
            || !output.is_char_boundary(end)
        {
            continue;
        }

        let label = &output[start..end];
        let linked = format!("[{label}]({})", reference.link);
        output.replace_range(start..end, &linked);
    }

    output
}

fn tokenize(text: &str) -> Vec<Token> {
    let mut tokens = Vec::new();

    for m in TOKEN_RE.find_iter(text) {
        let raw = m.as_str();
        if SECTION_NUMBER_RE.is_match(raw) {
            tokens.push(Token::SectionNumber {
                value: raw.to_lowercase(),
                start: m.start(),
                end: m.end(),
            });
            continue;
        }

        if let Some(caps) = DESIGNATOR_RE.captures(raw) {
            tokens.push(Token::Designator {
                value: caps[1].to_string(),
            });
            continue;
        }

        if raw.len() == 1 {
            let value = raw
                .chars()
                .next()
                .expect("punctuation should have one char");
            if matches!(value, ',' | ';' | '.' | ':') {
                tokens.push(Token::Punct { value });
                continue;
            }
        }

        tokens.push(Token::Word {
            value: raw.to_lowercase(),
        });
    }

    tokens
}

fn is_qualifier_keyword(token: &Token) -> bool {
    matches!(token, Token::Word { value } if QUALIFIER_KEYWORDS.contains(value.as_str()))
}

fn is_section_keyword(token: &Token) -> bool {
    matches!(token, Token::Word { value } if SECTION_KEYWORDS.contains(value.as_str()))
}

fn is_word(token: Option<&Token>, expected: &str) -> bool {
    matches!(token, Some(Token::Word { value }) if value == expected)
}

fn is_designator(token: Option<&Token>) -> bool {
    matches!(token, Some(Token::Designator { .. }))
}

fn is_section_number(token: &Token) -> bool {
    matches!(token, Token::SectionNumber { .. })
}

fn is_separator(token: Option<&Token>) -> bool {
    match token {
        Some(Token::Punct { value }) => *value == ',' || *value == ';',
        Some(Token::Word { value }) => SEPARATOR_WORDS.contains(value.as_str()),
        _ => false,
    }
}

struct ParsedReferences {
    references: Vec<SectionCrossReference>,
    next_index: usize,
}

fn parse_reference(tokens: &[Token], start_index: usize) -> Option<ParsedReferences> {
    let token = tokens.get(start_index)?;

    if is_qualifier_keyword(token) {
        let qualifier_chains = parse_qualifier_chain_list(tokens, start_index)?;
        let mut index = qualifier_chains.next_index;
        if !is_word(tokens.get(index), "of") {
            return None;
        }
        index += 1;

        let section_keyword = tokens.get(index)?;
        if !is_section_keyword(section_keyword) {
            return None;
        }

        let allow_multiple = matches!(section_keyword, Token::Word { value } if value == "sections" || value == "secs");
        let section_list = parse_section_list(tokens, index + 1, allow_multiple)?;
        let references = build_references(&section_list.items);
        return Some(ParsedReferences {
            references,
            next_index: section_list.next_index,
        });
    }

    if is_section_keyword(token) {
        let section_list = parse_section_list(tokens, start_index + 1, true)?;
        let references = build_references(&section_list.items);
        return Some(ParsedReferences {
            references,
            next_index: section_list.next_index,
        });
    }

    None
}

struct QualifierChainListResult {
    next_index: usize,
}

fn parse_qualifier_chain_list(
    tokens: &[Token],
    start_index: usize,
) -> Option<QualifierChainListResult> {
    let first_chain = parse_qualifier_chain(tokens, start_index)?;
    let mut index = first_chain.next_index;

    loop {
        let sep_index = match consume_separators(tokens, index) {
            Some(value) => value,
            None => break,
        };

        let next_token = tokens.get(sep_index)?;
        if !is_qualifier_keyword(next_token) {
            break;
        }

        let next_chain = parse_qualifier_chain(tokens, sep_index)?;
        index = next_chain.next_index;
    }

    Some(QualifierChainListResult { next_index: index })
}

struct QualifierChainResult {
    next_index: usize,
}

fn parse_qualifier_chain(tokens: &[Token], start_index: usize) -> Option<QualifierChainResult> {
    let qualifier = parse_qualifier(tokens, start_index)?;
    let mut index = qualifier.next_index;

    while is_word(tokens.get(index), "of") {
        let next_token = tokens.get(index + 1)?;
        if !is_qualifier_keyword(next_token) {
            break;
        }

        let next_qualifier = parse_qualifier(tokens, index + 1)?;
        index = next_qualifier.next_index;
    }

    Some(QualifierChainResult { next_index: index })
}

struct QualifierResult {
    next_index: usize,
}

fn parse_qualifier(tokens: &[Token], start_index: usize) -> Option<QualifierResult> {
    let token = tokens.get(start_index)?;
    if !is_qualifier_keyword(token) {
        return None;
    }

    let list = parse_designator_list(tokens, start_index + 1)?;
    Some(QualifierResult {
        next_index: list.next_index,
    })
}

struct DesignatorListResult {
    next_index: usize,
}

fn parse_designator_list(tokens: &[Token], start_index: usize) -> Option<DesignatorListResult> {
    let first = tokens.get(start_index);
    if !is_designator(first) {
        return None;
    }

    let mut index = start_index + 1;
    loop {
        let sep_index = match consume_separators(tokens, index) {
            Some(value) => value,
            None => break,
        };

        if !is_designator(tokens.get(sep_index)) {
            break;
        }

        index = sep_index + 1;
    }

    Some(DesignatorListResult { next_index: index })
}

struct SectionListResult {
    items: Vec<SectionTarget>,
    next_index: usize,
}

fn parse_section_list(
    tokens: &[Token],
    start_index: usize,
    allow_multiple: bool,
) -> Option<SectionListResult> {
    let first_item = parse_section_item(tokens, start_index)?;
    let mut items = vec![first_item.item];
    let mut index = first_item.next_index;

    if !allow_multiple {
        return Some(SectionListResult {
            items,
            next_index: index,
        });
    }

    loop {
        let sep_index = match consume_separators(tokens, index) {
            Some(value) => value,
            None => break,
        };

        let mut next_index = sep_index;
        if matches!(tokens.get(next_index), Some(token) if is_section_keyword(token)) {
            next_index += 1;
        }

        let next_item = match parse_section_item(tokens, next_index) {
            Some(value) => value,
            None => break,
        };

        items.push(next_item.item);
        index = next_item.next_index;
    }

    Some(SectionListResult {
        items,
        next_index: index,
    })
}

struct SectionItemResult {
    item: SectionTarget,
    next_index: usize,
}

fn parse_section_item(tokens: &[Token], start_index: usize) -> Option<SectionItemResult> {
    let token = tokens.get(start_index)?;
    let (value, start, end) = match token {
        Token::SectionNumber { value, start, end } => (value.clone(), *start, *end),
        _ => return None,
    };

    let mut index = start_index + 1;
    let start_mention = SectionMention {
        section: value,
        offset: start,
        length: end - start,
    };

    if is_word(tokens.get(index), "to") {
        let end_token = tokens.get(index + 1)?;
        let (end_value, end_start, end_end) = match end_token {
            Token::SectionNumber { value, start, end } => (value.clone(), *start, *end),
            _ => return None,
        };

        index += 2;

        let mut inclusive = false;
        if matches!(tokens.get(index), Some(Token::Punct { value: ',' })) {
            if is_word(tokens.get(index + 1), "inclusive") {
                inclusive = true;
                index += 2;
            }
        } else if is_word(tokens.get(index), "inclusive") {
            inclusive = true;
            index += 1;
        }

        let end_mention = SectionMention {
            section: end_value,
            offset: end_start,
            length: end_end - end_start,
        };

        return Some(SectionItemResult {
            item: SectionTarget::Range {
                start: start_mention,
                end: end_mention,
                inclusive,
            },
            next_index: index,
        });
    }

    Some(SectionItemResult {
        item: SectionTarget::Section {
            mention: start_mention,
        },
        next_index: index,
    })
}

fn consume_separators(tokens: &[Token], start_index: usize) -> Option<usize> {
    let mut index = start_index;
    let mut consumed = false;

    while is_separator(tokens.get(index)) {
        consumed = true;
        index += 1;
    }

    if consumed {
        Some(index)
    } else {
        None
    }
}

fn build_references(items: &[SectionTarget]) -> Vec<SectionCrossReference> {
    let mut references = Vec::new();

    for item in items {
        match item {
            SectionTarget::Section { mention } => references.push(build_reference(mention)),
            SectionTarget::Range { start, end, .. } => {
                references.push(build_reference(start));
                references.push(build_reference(end));
            }
        }
    }

    references
}

fn dedupe_references(references: Vec<SectionCrossReference>) -> Vec<SectionCrossReference> {
    let mut seen = HashSet::new();
    let mut deduped = Vec::new();

    for reference in references {
        let key = format!(
            "{}:{}:{}:{}",
            reference.section, reference.offset, reference.length, reference.link
        );
        if seen.insert(key) {
            deduped.push(reference);
        }
    }

    deduped
}

fn build_reference(mention: &SectionMention) -> SectionCrossReference {
    SectionCrossReference {
        section: mention.section.clone(),
        offset: mention.offset,
        length: mention.length,
        link: format!("/statutes/section/{}", mention.section),
    }
}
